src/hooks/useURLState.ts

```diff
@@ -0,0 +1,84 @@
+import React from 'react';
+import { useHistory, useLocation } from 'react-router-dom';
+import { filtersFromQuery } from '../attribute-filters';
+import { AttributeList, Filters } from '../components/filters/filter.types';
+import { useQueryParams } from './useQueryParams';
+
+interface UseURLStateHook {
+  defaultQuery?: string;
+  attributes: AttributeList;
+}
+
+const QUERY_PARAM_KEY = 'q';
+const TENANT_PARAM_KEY = 'tenant';
+const SHOW_RESOURCES_PARAM_KEY = 'showResources';
+const DEFAULT_TENANT = 'application';
+const DEFAULT_SHOW_RESOURCES = '0';
+const DEFAULT_QUERY = '{ log_type =~ ".+" } | json';
+
+export const useURLState = ({
+  defaultQuery = DEFAULT_QUERY,
+  attributes,
+}: UseURLStateHook) => {
+  const queryParams = useQueryParams();
+  const history = useHistory();
+  const location = useLocation();
+
+  const initialQuery = queryParams.get(QUERY_PARAM_KEY) ?? defaultQuery;
+  const initialTenant = queryParams.get(TENANT_PARAM_KEY) ?? DEFAULT_TENANT;
+
+  const [query, setQuery] = React.useState(initialQuery);
+  const [tenant, setTenant] = React.useState(initialTenant);
+  const [filters, setFilters] = React.useState<Filters | undefined>({});
+  const [areResourcesShown, setAreResourcesShown] = React.useState(false);
+
+  const setTenantInURL = (tenant: string) => {
+    queryParams.set(TENANT_PARAM_KEY, tenant);
+    history.push(`${location.pathname}?${queryParams.toString()}`);
+  };
+
+  const setShowResourcesInURL = (showResources: boolean) => {
+    queryParams.set(SHOW_RESOURCES_PARAM_KEY, showResources ? '1' : '0');
+    history.push(`${location.pathname}?${queryParams.toString()}`);
+  };
+
+  const setQueryInURL = (newQuery: string) => {
+    queryParams.set(QUERY_PARAM_KEY, newQuery);
+    history.push(`${location.pathname}?${queryParams.toString()}`);
+  };
+
+  const setValuesFromParams = (newQueryParams: URLSearchParams) => {
```

Comment:
For my comment https://github.com/openshift/logging-view-plugin/pull/27#pullrequestreview-1133991541, rather than just renaming the parameter, I was wondering if you could avoid passing in a parameter at all and use `queryParams` here? Or maybe `queryParams` won't yet be set when the function is first called?

Taking it a bit further, would something like this work?
```
useEffect(() => {
  const queryValue = queryParams.get(QUERY_PARAM_KEY) ?? initialQuery;
  const tenantValue = queryParams.get(TENANT_PARAM_KEY) ?? DEFAULT_TENANT;
  const showResourcesValue =
    queryParams.get(SHOW_RESOURCES_PARAM_KEY) ?? DEFAULT_SHOW_RESOURCES;

  setQuery(queryValue);
  setTenant(tenantValue);
  setAreResourcesShown(showResourcesValue === '1');
  setFilters(filtersFromQuery({ query: queryValue, attributes }));
}, [queryParams])
```
And then you wouldn't need the other two `useEffect`s at all.

Reply:
Yes, having the useEffect listening into the history rather than the `queryValue` allowed that the text input didn't loose the caret position. 

But in fact this was an issue of the input and shouldn't be solved in the useURLState hook. I changed it to use just one useEffect and adjusted the input proxying the prop in an internal state so the caret position is kept while the url parameter changes.